### PR TITLE
Merge ultrasonic-poller into master

### PR DESCRIPTION
**Pull request by ddrobner**
Take a look at the logic in the task for UltrasonicPoller, I'm pretty sure it's good.

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java [hunk 1]
```diff
@@ -12,13 +14,22 @@
...
         super(1, TimeUnit.SECONDS);
         
         this.beltController = new WPI_TalonSRX(RobotMap.BELT_TALON_PORT);
         this.rollerController = new WPI_TalonSRX(RobotMap.ROLLER_TALON_PORT);
         this.feederUltrasonic = new Ultrasonic(RobotMap.ULTRASONIC_PORTS[0], RobotMap.ULTRASONIC_PORTS[1]);
+        
+        this.ballDistance = 300;
+        this.distanceTuner = new SmartDashboardTuner("Ball Distance", this.ballDistance, d -> this.ballDistance = d);
+
+        this.poller = new UltrasonicPoller(feederUltrasonic, 3000, this.ballDistance);
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java]: This won't work how you want it to, because the `ballDistance` value is only passed into the UltrasonicPoller constructor once, so any changes you make using the tuner will not actually do anything other than change the field value within `FeederImpl`.

If you put the tuner inside the poller then it'll probably do what you're expecting.

File: src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java [hunk 2]
```diff
@@ -44,8 +55,10 @@ public void defineResources(){
     }
 
     public void unload() {
-        this.reverseFeeder();
-        // TODO implement sensor logic
+        while(!this.poller.checkUnloaded()){
+            this.reverseFeeder();
+        }
+        this.stop();
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java:59]: Who are all the possible callers of `unload`? Is it safe to put a while-loop here, especially one without a `sleep` in it? If whatever calls this is going to do so periodically then the while-loop can be turned into a simple if-statement.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 3]
```diff
@@ -9,22 +10,50 @@
...
+        super(rate, TimeUnit.MILLISECONDS); // 4Hz seems reasonable
         this.ultrasonicSensor = sensor;
+        this.time = time;
+        this.distance = distance;
+        this.noBalls = false;
+
+    }
+
+    UltrasonicPoller(Ultrasonic sensor, long time, double distance) {
+        this(sensor, time, distance, 25);
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:35]: What's the reason for having both of these constructors?

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 4]
```diff
@@ -9,22 +10,50 @@
 class UltrasonicPoller extends RepeatingPooledSubsystem {
 
     Ultrasonic ultrasonicSensor;
+    StopWatch timer = new StopWatch();
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Pulling in a whole Apache Commons dependency just for this is way, way unnecessary.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 5]
```diff
@@ -9,22 +10,50 @@
...
     @Override
     public void task(){
-        // Placholder
-        return;
+        if(!this.timer.isStarted()){
+            this.timer.start();
+        }
+        this.currentRange = this.ultrasonicSensor.getRangeMM();
+        if((this.currentRange > this.distance) && (this.timer.getTime() > this.time)) {
+            this.noBalls = true;
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:65]: `noBalls` is initially `false` and becomes `true` under this condition. When can it transition back to being `false`?

**Review by andrewazores — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java (hunk 2 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java:59]: Yeah that's true. I didn't think about that.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 3 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:35]: Default refresh rate

Comment (andrewazores) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:35]: I can see that, but why do you need to have a constructor for a default rate and another that allows it to be specified by whoever's constructing it?

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 4 above)
Comment (andrewazores) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: How about using a `long` and `System.nanoTime` instead of this stopwatch? Test out `long last = System.nanoTime(); Thread.sleep(1000); long current = System.nanoTime(); long elapsed = current - last;` and see how this might be useful.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 3 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:35]: True. Considering this was potentially going to be merged into Yggdrasil, I thought that I might as well do it semi-right the first time.

Comment (andrewazores) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:35]: Just keep it clean and simple for now. There's going to be other changes needed before upstreaming anyway, so don't try to plan too far in advance.

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java:35]: Ok, I'll just leave it though, no point in removing it if it's already there.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 4 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: I changed it to use System.nanoTime. Take a look at it, I'm pretty sure it should work.

File: src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java (hunk 2 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/FeederImpl.java:59]: AFAIK the controllers only call the method once (when you map the controls), so what's calling it isn't doing it periodically. I'll add a sleep to it for a few ms.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 6]
```diff
@@ -10,21 +10,58 @@
...
+        return this.noBalls;
+    }
+
     @Override
     public void task(){
-        // Placholder
-        return;
+        this.currentRange = this.ultrasonicSensor.getRangeMM();
+        this.startTime = System.nanoTime();
+        if((this.currentRange > this.distance) && (((System.nanoTime() - this.startTime) / 100000) > this.time)) {
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Your two calls to `System.nanoTime()` happen back-to-back, so there's almost certainly no chance that the time threshold will have elapsed by the second evaluation. You need to store the `startTime` timestamp only on the "edges" of the state transition - currently you're setting it every time the task evaluates.

```
[X]  [X]  [ ]  [ ]  [ ]  [ ]      [ ]  [ ]  [ ]  [X]
           ^                 ...   ^              ^
          mark                   threshold       reset
```

**Review by andrewazores — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 6 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Yeah that's true. I made some modifications and didn't stop to reconsider the implications of them.

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Lol wait, I just realized I made a major logic error. What I really want to do is only get the start time on the first run of the task (that's probably what I intended last night, I just wasn't thinking), then compare the difference and reset the start time only if a ball is detected.

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: The reason this happened is because I noticed that because startTime has an initial value of 0, it will always set noBalls to true (because System.nanoTime - 0 = System.nanotime, which will be much bigger than whatever the time threshold is), so I said: "Ok, let's set the start time to avoid this edge case from happening", so I set the start time at the beginning of the task. What I neglected is that the task is keep on looping while it's enabled, so I didn't think about how the time will get reset on every run, regardless of what's happening.

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: The two commits I just made should resolve this issue, but I feel as if the way I went about it was, although perfectly functional and doesn't hit performance at all (literally just a single comparison), horrible stylistically, and too 'bloated'.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 7]
```diff
@@ -10,21 +10,63 @@
...
     @Override
     public void task(){
-        // Placholder
-        return;
+        this.currentRange = this.ultrasonicSensor.getRangeMM();
+        if(this.count == 0){ // TODO look into a better way of doing this
+            this.startTime = System.nanoTime(); // I know about a do while, but this isn't contained within a while loop 
+            this.count = 1;
+        }
+        if((this.currentRange > this.distance) && (((System.nanoTime() - this.startTime) / 100000) > this.time)) {
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Rather than dividing by `100000` (did you mean to use one hundred thousand here?), for better readability or ease of changing it in the future, please use something like `TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - this.startTime()) > time`

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java [hunk 8]
```diff
@@ -10,21 +10,63 @@
...
+    public boolean checkUnloaded(){
+        return this.noBalls;
+    }
+
     @Override
     public void task(){
-        // Placholder
-        return;
+        this.currentRange = this.ultrasonicSensor.getRangeMM();
+        if(this.count == 0){ // TODO look into a better way of doing this
```
Comment (andrewazores) [inline src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Okay, this is getting a little messy and there are definitely ways to streamline this and keep fewer redundant fields, but I think at this point your best bet is probably going to be to just test this out on real hardware and see how it behaves. That'll help you get a good mental model of what information is available and how you can best model it.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 7 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: I meant one-million. Irrelevant now anyways since I'll just use toMillis.

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 8 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Just figured out how I could rewrite it
```
do {
set the initial time
} while the initial time is zero
// rest of the task
```

Comment (andrewazores) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: That sounds a lot like

```
if initial time is zero
  set the initial time
```

no?

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Dude how did I miss that. It's been there in front of me this entire time but I was too blind to see it.

**Review by andrewazores — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java (hunk 8 above)
Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: Granted - both of these approaches will cause an issue if the system is stopped and then started again (but that issue is insignificant because I doubt that the output of the poller will be checked within the 20ms timeframe that noBalls is an incorrect value), but how likely are we to do that. Now that I think about it, is it even necessary to do this? If noBalls holds an incorrect value when the subsystem is started, it's only for 20ms until the loop runs again, where if it detects a ball it will set noBalls to false. However, that does cause issues where, say the threshold is 3 seconds and there are no balls for two seconds, it will return a true value, even though it should return false.

Comment (andrewazores) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: I'm not really following your explanation, but yea, there are still edge cases you might need to iron out. I think what you have here is workable enough to test on real hardware and tinker with it from there though, so I'm going to merge this and you can open a new PR with any refinements.

Comment (ddrobner) [reply src/main/java/frc/robot/subsystems/ball/feeder/UltrasonicPoller.java]: That's implying the robot will actually be together in time for me to do any tinkering
